Che workspace like Python, PHP ... deployment fails

### General information

  * Minishift version: master
  * OS: Linux / macOS / Windows
  * Hypervisor: KVM / xhyve / Hyper-V / VirtualBox


### Steps to reproduce

  1. make integration GODOG_OPTS="-tags addon-che"
  2. 
  3. 
  4. 

### Expected
Should go through

### Actual
Fails - https://ci.centos.org/job/minishift-nightly-centos/328/consoleFull
           https://ci.centos.org/job/minishift-nightly-centos/329/consoleFull

### Logs

You can start Minishift with `minishift start --show-libmachine-logs -v5` to collect logs.
Please consider posting this on http://gist.github.com/ and post the link in the issue.

Even with higher timeout the workspaces are not able to start consistently. I've added more debug outputs and found out that the first workspace is not properly deleted. Even though Che API confirms it is deleted, the `oc` says something slightly different (Vert.X workspace is running for 40 minutes even though it was deleted 3 workspace deployments ago):

```
12:54:41       | CentOS WildFly Swarm | https://github.com/wildfly-swarm-openshiftio-boosters/wfswarm-rest-http |
12:54:41         time limit (10m) exceeded: workspace did not get into state 'RUNNING', current state: '{STOPPED}'

12:55:44 NAME                                                    READY     STATUS    RESTARTS   AGE
12:55:44 che-1-xfrfl                                             1/1       Running   0          44m
12:55:44 workspacen79p9px0dkcauptp.dockerimage-bb48dcd64-6k5gc   1/1       Running   0          40m
```

The Che logs shows that the error occurs probably during the stop, maybe a deployment config then is still present and thus killed node is redeployed after the deletion? Here's the log:

```
12:33:23 2019-01-30 12:20:35,731[nio-8080-exec-5]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 473]   - Workspace 'che/vert.x-stack-test' with id 'workspacen79p9px0dkcauptp' is stopping by user 'che'
12:33:23 2019-01-30 12:20:46,801[aceSharedPool-1]  [ERROR] [o.e.c.a.w.s.WorkspaceRuntimes 536]   - Error occurred during stopping of runtime 'workspacen79p9px0dkcauptp:default:che' by user 'che'. Error: Error(s) occurs while cleaning up the namespace. Operation: [patch]  for kind: [Deployment]  with name: [workspacen79p9px0dkcauptp.dockerimage]  in namespace: [mini-che]  failed.
12:33:23 org.eclipse.che.api.workspace.server.spi.InfrastructureException: Error(s) occurs while cleaning up the namespace. Operation: [patch]  for kind: [Deployment]  with name: [workspacen79p9px0dkcauptp.dockerimage]  in namespace: [mini-che]  failed.
12:33:23 	at org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace.doRemove(KubernetesNamespace.java:192)
12:33:23 	at org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftProject.cleanUp(OpenShiftProject.java:100)
12:33:23 	at org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesInternalRuntime.internalStop(KubernetesInternalRuntime.java:619)
12:33:23 	at org.eclipse.che.api.workspace.server.spi.InternalRuntime.stop(InternalRuntime.java:177)
12:33:23 	at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$StopRuntimeTask.run(WorkspaceRuntimes.java:506)
12:33:23 	at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:38)
12:33:23 	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1626)
12:33:23 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
12:33:23 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
12:33:23 	at java.lang.Thread.run(Thread.java:748)
12:33:23 2019-01-30 12:20:51,497[nio-8080-exec-6]  [INFO ] [o.e.c.a.w.s.WorkspaceManager 274]    - Workspace 'workspacen79p9px0dkcauptp' removed by user 'che'
```

--- a/addons/che/templates/che-server-template.yaml
+++ b/addons/che/templates/che-server-template.yaml
@@ -96,8 +96,8 @@ objects:
             value: "openshift"
           - name: CHE_INFRA_KUBERNETES_BOOTSTRAPPER_BINARY__URL
             value: "${PROTOCOL}://che-${NAMESPACE}.${ROUTING_SUFFIX}/agent-binaries/linux_amd64/bootstrapper/bootstrapper"
-          - name: CHE_INFRA_KUBERNETES_MACHINE__START__TIMEOUT__MIN
-            value: "5"
+          - name: CHE_INFRA_KUBERNETES_WORKSPACE__START__TIMEOUT__MIN
+            value: "10"
           - name: CHE_INFRA_KUBERNETES_MASTER__URL
             value: "${CHE_INFRA_KUBERNETES_MASTER__URL}"
           - name: CHE_INFRA_OPENSHIFT_PROJECT

--- a/test/integration/che/cheAPI.go
+++ b/test/integration/che/cheAPI.go
@@ -451,8 +451,11 @@ func (c *CheAPI) GetWorkspaceStatusByID(workspaceID string) (WorkspaceStatus, er
 
 	var workspaceStatusObj WorkspaceStatus
 	err = json.Unmarshal(workspaceDataJSON, &workspaceStatusObj)
+	if err != nil {
+		return workspaceStatusObj, fmt.Errorf("error unmarshaling JSON: '%s' into '%v', error: %v", workspaceDataJSON, workspaceStatusObj, err)
+	}
 
-	return workspaceStatusObj, err
+	return workspaceStatusObj, nil
 }
 
 //CheckWorkspaceDeletion checks if the workspace at workspaceID is deleted

--- a/test/integration/features/addons/addon-che-with-anyuid.feature
+++ b/test/integration/features/addons/addon-che-with-anyuid.feature
@@ -23,7 +23,7 @@ Feature: Che add-on in combination with anyuid addon.
   Scenario: Che is ready
     Given Minishift has state "Running"
      When executing "oc project mini-che" succeeds
-     Then service "che" rollout successfully within "5m"
+     Then service "che" rollout successfully within "10m"
 
   Scenario: Che API is accessible
      When user tries to get the che api endpoint
@@ -34,6 +34,10 @@ Feature: Che add-on in combination with anyuid addon.
      Then the stacks should not be empty
 
   Scenario Outline: User starts workspace, imports projects, checks run commands
+    # workaround for https://github.com/eclipse/che/issues/12569
+    Given executing "oc delete all -l che.workspace_id" retrying 10 times with wait period of "10s"
+      And executing "oc delete pvc -l che.workspace_id" retrying 10 times with wait period of "10s"
+
      When starting a workspace with stack "<stack>" succeeds within "15m"
      Then workspace should have state "RUNNING"
 

--- a/test/integration/features/addons/addon-che.feature
+++ b/test/integration/features/addons/addon-che.feature
@@ -18,7 +18,7 @@ Feature: Che add-on
   Scenario: Che is ready
     Given Minishift has state "Running"
      When executing "oc project mini-che" succeeds
-     Then service "che" rollout successfully within "5m"
+     Then service "che" rollout successfully within "10m"
 
   Scenario: Che API is accessible
      When user tries to get the che api endpoint
@@ -29,6 +29,10 @@ Feature: Che add-on
      Then the stacks should not be empty
 
   Scenario Outline: User starts workspace, imports projects, checks run commands
+    # workaround for https://github.com/eclipse/che/issues/12569
+    Given executing "oc delete all -l che.workspace_id" retrying 10 times with wait period of "10s"
+      And executing "oc delete pvc -l che.workspace_id" retrying 10 times with wait period of "10s"
+
      When starting a workspace with stack "<stack>" succeeds within "15m"
      Then workspace should have state "RUNNING"
 

--- a/test/integration/features/addons/cmd-addons.feature
+++ b/test/integration/features/addons/cmd-addons.feature
@@ -189,10 +189,6 @@ Feature: Addons command and its subcommands
      Then stdout should contain "Removing addon 'admin-user'"
       And stdout should contain "admin user deleted"
 
-  Scenario: Applying already applied add-on
-  The result would differ from add-on to add-on, but will probably fail.
-  Thus this case is not being tested.
-
   Scenario: Stopping Minishift
     Given Minishift has state "Running"
      When executing "minishift stop" succeeds

--- a/test/integration/testsuite/testsuite.go
+++ b/test/integration/testsuite/testsuite.go
@@ -264,18 +264,12 @@ func FeatureContext(s *godog.Suite) {
 		writeToFileInTestDirSucceeds)
 	s.Step(`^file "([^"]*)" should match text "([^"]*)" succeeds$`,
 		readFileForTextMatchSucceeds)
-	// Prototyping and debugging
-	// please do not use in production
+	// Prototyping and debugging steps
+	// do not use in production extensively
 	s.Step(`^user (?:waits|waited) for "(\d+(?:ms|s|m))"$`,
-		func(duration string) error {
-			timeDuration, err := time.ParseDuration(duration)
-			if err != nil {
-				return err
-			}
-
-			time.Sleep(timeDuration)
-			return nil
-		})
+		waitForDuration)
+	s.Step(`^echo (stdout|stderr|exitcode)$`,
+		echoLastCommandOutput)
 
 	s.BeforeSuite(func() {
 		testDir, testDefaultHome = setupTestDirectory()
@@ -657,6 +651,21 @@ func ExecutingOcCommandSucceedsOrFails(command string, expectedResult string) er
 
 func ExecutingMinishiftCommandSucceedsOrFails(command string, expectedResult string) error {
 	return succeedsOrFails(MinishiftInstance.ExecutingMinishiftCommand, command, expectedResult)
+}
+
+func waitForDuration(duration string) error {
+	timeDuration, err := time.ParseDuration(duration)
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(timeDuration)
+	return nil
+}
+
+func echoLastCommandOutput(commandField string) error {
+	fmt.Println(selectFieldFromLastOutput(commandField))
+	return nil
 }
 
 func succeedsOrFails(execute commandRunner, command string, expectedResult string) error {

